pkg/master/controller.go

```diff
@@ -155,15 +174,14 @@ func (c *Controller) CreateMasterServiceIfNeeded(serviceName string, serviceIP n
 			Labels:    map[string]string{"provider": "kubernetes", "component": "apiserver"},
 		},
 		Spec: api.ServiceSpec{
-			Ports: []api.ServicePort{{Port: servicePort, Protocol: api.ProtocolTCP, TargetPort: util.NewIntOrStringFromInt(servicePort)}},
+			Ports: ports,
 			// maintained by this code, not by the pod selector
 			Selector:        nil,
 			ClusterIP:       serviceIP.String(),
```

Comment:
NodePort is only half of the solution for that, it doesn't give you a
single IP:port, it gives you port X on every node. You still need some sort
of load balancer, the setup of which depends heavily on the environment.

On Fri, Sep 25, 2015 at 7:05 AM, Timothy St. Clair <notifications@github.com
> wrote:

> In pkg/master/controller.go
> <https://github.com/kubernetes/kubernetes/pull/13978#discussion_r40432465>
> :
>
> > @@ -155,15 +174,14 @@ func (c *Controller) CreateMasterServiceIfNeeded(serviceName string, serviceIP n
> >  			Labels:    map[string]string{"provider": "kubernetes", "component": "apiserver"},
> >  		},
> >  		Spec: api.ServiceSpec{
> > -			Ports: []api.ServicePort{{Port: servicePort, Protocol: api.ProtocolTCP, TargetPort: util.NewIntOrStringFromInt(servicePort)}},
> > +			Ports: ports,
> >  			// maintained by this code, not by the pod selector
> >  			Selector:        nil,
> >  			ClusterIP:       serviceIP.String(),
>
> @thockin <https://github.com/thockin> we would like a single cluster
> IP:port for a cluster that has mulitple api-servers. Currently in our HA
> configurations we stand up an eternal HA-Proxy and VIP which is managed by
> pacemaker. This VIP provides our single load-balanced endpoint into the
> cluster itself.
>
> The experimentation with NodePort was to see if we could expose the
> load-balanced introspective service when we have multiple api-servers. All
> that being said, we recognize that @bprashanth
> <https://github.com/bprashanth> ' s new method might be cleaner.
>
> Hope this clarifies the objective.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/13978/files#r40432465>.
>